cpp-ch/local-engine/Operator/ExpandTransform.cpp

```diff
@@ -68,43 +67,36 @@ ExpandTransform::Status ExpandTransform::prepare()
 void ExpandTransform::work()
 {
     assert(expanded_chunks.empty());
-    size_t agg_cols_size = aggregating_expressions_columns.size();
-    for (int set_id = 0; static_cast<size_t>(set_id) < grouping_sets.size(); ++set_id)
+    for (const auto & project_exprs: project_set_exprs)
     {
-        const auto & sets = grouping_sets[set_id];
-        DB::Columns cols;
         const auto & original_cols = input_chunk.getColumns();
-        for (size_t i = 0; i < original_cols.size(); ++i)
+        size_t rows = input_chunk.getNumRows();
+        DB::Columns cols;
+        for (const auto & project_expr : project_exprs)
         {
-            const auto & original_col = original_cols[i];
-            size_t rows = original_col->size();
-            if (i < agg_cols_size)
-            {
-                cols.push_back(original_col);
-                continue;
-            }
-            // the output columns should all be nullable.
-            if (!sets.contains(i))
+            if (project_expr.kind == EXPAND_FIELD_KIND_SELECTION)
             {
-                auto null_map = DB::ColumnUInt8::create(rows, 1);
-                auto col = DB::ColumnNullable::create(original_col, std::move(null_map));
-                cols.push_back(std::move(col));
+                cols.push_back(original_cols[project_expr.field.get<Int32>()]);
             }
             else
             {
-                if (original_col->isNullable())
-                    cols.push_back(original_col);
+                if (project_expr.field.isNull())
+                {
+                    // Add null column
+                    auto null_map = DB::ColumnUInt8::create(rows, 1);
+                    auto nested_type = DB::removeNullable(project_expr.type);
+                    auto col = DB::ColumnNullable::create(nested_type->createColumn()->cloneResized(rows), std::move(null_map));
+                    cols.push_back(std::move(col));
+                }
                 else
                 {
-                    auto null_map = DB::ColumnUInt8::create(rows, 0);
-                    auto col = DB::ColumnNullable::create(original_col, std::move(null_map));
+                    // Add constant column: gid, gpos, etc.
+                    auto col = project_expr.type->createColumnConst(rows, project_expr.field);
```

Comment:
Do we realy need to make a projection for the literal ahead , if we add the const gid column by this way?

Or Do we have the cases that a literal that is not the gid in the substait rel?

Reply:
This implementation is more simple. A literal in expand projection list should only be gid or gpos.